### PR TITLE
Allow `EndpointConfig::configureTypes()` to customize object, interface and union type code generation

### PR DESCRIPTION
**Pull request by spawnia**
https://github.com/spawnia/sailor/discussions/86

### CODE REVIEW

File: examples/custom-types/sailor.php [hunk 1]
```diff
@@ -42,6 +43,15 @@ public function makeClient(): Client
                     'withCustomEnum' => CustomEnum::B,
                 ],
             ]);
+            $mockClient->responseMocks[] = static fn (string $query, ?\stdClass $variables): Response => Response::fromStdClass((object) [
```
Comment (spawnia) [inline examples/custom-types/sailor.php:46]: It appears that `$mockClient` is instantiated every multiple times, thus the consecutive tests across different operations in `examples/custom-types/src/test.php` do not work correctly.

**Comment by spawnia** [comment]
@morloderex please test and review.

**Comment by morloderex** [comment]
@spawnia I am sorry it has taken me a bit of time to get to this one again. 

I will be trying it within a few days and report back :)

**Comment by morloderex** [comment]
Hey @spawnia 

I have now tried this out, and it seems to sort of work. But only if you do a query that has the custom type defined once.

I don't know if this is something we can do anything about sense it seems to be given an exception in the `operationStack::selection()` method. It seems you are only working with the fields and not the entire object?

For instance the fallowing query fails to compile:

```graphql
query account($id: ID!) {
    account(accountId: $id) {
        balances {
            __typename
            available {
                value
                currency
            }
            pending {
                __typename
                value
                currency
            }
        }
    }
}
``` 

With the introspection as fallowing: 
```
type AccountBalances {
  "Balance composed of booked, pending transactions and rolling reserve used known at the time of calculation."
  available: Amount!

  "Balance just composed of pending transactions."
  pending: Amount!

  "Balance just composed of booked transactions."
  booked: Amount!

  "Balance just composed of rolling reserve used in booked transactions."
  reserved: Amount!
}

type Account {
    balances: AccountBalances
}
```

And the custom Type configured for the Amount type: 
```php
<?php declare(strict_types=1);

namespace LaravelSwan\development;

use GraphQL\Type\Definition\NamedType;
use GraphQL\Type\Definition\Type;
use GraphQL\Type\Schema;
use Money\Money;
use Nette\PhpGenerator\ClassType;
use Nette\PhpGenerator\Method;
use Spawnia\Sailor\Convert\GeneratesTypeConverter;
use Spawnia\Sailor\EndpointConfig;
use Spawnia\Sailor\Type\InputTypeConfig;
use Spawnia\Sailor\Type\OutputTypeConfig;
use Spawnia\Sailor\Type\TypeConfig;

final class AmountTypeConfig implements TypeConfig, OutputTypeConfig, InputTypeConfig
{
    use GeneratesTypeConverter;

    private EndpointConfig $endpointConfig;

    /**
     * @var Type&NamedType
     */
    private Type $type;
    private Schema $schema;

    /**
     * @param Type&NamedType $type
     */
    public function __construct(EndpointConfig $endpointConfig, Schema $schema, Type $type)
    {
        $this->endpointConfig = $endpointConfig;
        $this->type = $type;
        $this->schema = $schema;
    }

    public function typeConverter(): string
    {
        return $this->typeConverterClassName($this->type, $this->endpointConfig);
    }


    public function outputTypeReference(): string
    {
        return '\\' . Money::class;
    }

    public function generateClasses(): iterable
    {
        yield $this->makeTypeConverter($this->type, $this->endpointConfig);
    }

    protected function decorateTypeConverterClass(Type $type, ClassType $class, Method $fromGraphQL, Method $toGraphQL): ClassType
    {
        $class->getNamespace()->addUse(\Money\Parser\DecimalMoneyParser::class);
        $class->getNamespace()->addUse(\Money\Formatter\DecimalMoneyFormatter::class);
        $class->getNamespace()->addUse(\Money\Currencies\ISOCurrencies::class);
        $class->getNamespace()->addUse(\Money\Currency::class);
        $class->getNamespace()->addUse(\stdClass::class);
        $class->getNamespace()->addUse(\InvalidArgumentException::class);
        $class->getNamespace()->addUse(\Money\Money::class);

        $returnType = $class->getNamespace()->simplifyType(\Money\Money::class);

        $customObjectClass = Money::class;

        $fromGraphQL->setReturnType($customObjectClass);
        $fromGraphQL->setBody(
            <<<PHP
            if (! \$value instanceof stdClass) {
                throw new InvalidArgumentException('Expected stdClass, got: '.gettype(\$value));
            }

            \$arrayValue = (array) \$value;

            if (! array_key_exists('value', \$arrayValue) || ! array_key_exists( 'currency', \$arrayValue)) {
                throw new InvalidArgumentException('Did not find expected property value.');
            }

            return (new DecimalMoneyParser(new ISOCurrencies()))->parse(\$arrayValue['value'], new Currency(\$arrayValue['currency']));
            PHP
        );

        $toGraphQL->setBody(
            <<<PHP
                if (! \$value instanceof {$returnType}) {
                    throw new InvalidArgumentException('Expected instanceof {$returnType}, got: '.gettype(\$value));
                }

                return (object) ['value' => (new DecimalMoneyFormatter(new ISOCurrencies()))->format(\$value), 'currency' => \$value->getCurrency()->getCode()];
                PHP
        );

        return $class;
    }

    public function inputTypeReference(): string
    {
        return '\\' . Money::class;
    }
}
```

**Comment by spawnia** [comment]
@morloderex I could reproduce the issue you were seeing and fixed it, please try again with the latest changes I made.

**Comment by morloderex** [comment]
@spawnia I can confirm that your changes fixed the issue i was experiencing regarding the same custom type multiple times. 👍

**Pull request merged**